useeffect-and-calling-apis/useeffect-after-render.md

```diff
@@ -0,0 +1,542 @@
+
+##### !end-question
+##### !options
+
+* `currentTime`
+* `isEnabled`
+* `Timer`
+* None of the above
+
+##### !end-options
+##### !answer
+
+* `isEnabled`
+
+##### !end-answer
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<!-- Question 3 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: checkbox
+* id: d19f14f4
+* title: useEffect After Render
+##### !question
+
+Consider this code:
+
+```js
+import { useState, useEffect } from 'react';
+
+const Timer = () => {
+  const [currentTime, setCurrentTime] = useState(0);
+  const [isEnabled, setIsEnabled] = useState(false);
+
+  useEffect(() => {
+
+  }, [isEnabled, currentTime]);
+}
+
+```
+
+`useEffect` will run every time certain value(s) update. Which values will cause `useEffect` to run?
+
+##### !end-question
+##### !options
+
+* `currentTime`
+* `isEnabled`
+* `Timer`
+* None of the above
+
+##### !end-options
+##### !answer
+
+* `currentTime`
+* `isEnabled`
+
+##### !end-answer
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<!-- Question 4 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: checkbox
+* id: 4e323978
+* title: useEffect After Render
+##### !question
+
+Consider this code:
+
+```js
+import { useState, useEffect } from 'react';
+
+const Timer = () => {
+  const [currentTime, setCurrentTime] = useState(0);
+  const [isEnabled, setIsEnabled] = useState(false);
+
+  useEffect(() => {
+
+  }, []);
+}
+
+```
+
+`useEffect` will run every time certain value(s) update. Which values will cause `useEffect` to run?
+
+##### !end-question
+##### !options
+
+* `currentTime`
+* `isEnabled`
+* `Timer`
+* None of the above
+
+##### !end-options
+##### !answer
+
+* None of the above
+
+##### !end-answer
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+### !callout-danger
+
+## Our Examples Don't Utilize a Cleanup Function
+
+On the subject of `useEffect` syntax, the `useEffect` hook also supports defining a _clean-up_ function. "Clean-up functions" describe functions responsible for "cleaning up potential side-effects," such as closing database connections.
+
+<br/>
+
+In the examples in this curriculum, there will not be many or any examples that use the cleanup function meaningfully.
+
+<details>
+
+<summary>For those curious, here's an example of defining a clean-up function.</summary>
+
+In React, clean-up functions will run during `componentWillUnmount`. To define a clean-up function, we must _return_ the clean-up function inside our `useEffect` callback function.
+
+<!-- prettier-ignore-start -->
+```js
+import { useEffect } from 'react';
+
+function App() {
+
+  useEffect(() => {
+    console.log('I\'m in useEffect!');
+    console.log('This will be called whenever an instance of this component mounts');
+
+    return () => { console.log('useEffect calls the returned function during componentWillUnmount') };
+  }, []);
+
+  return (<h1>Demo of Clean-up Function Syntax</h1>);
+}
```

Comment:
the current example doesn't really show the unmount behavior occurring, so I fleshed this out a bit
```suggestion
import { useState, useEffect } from 'react';

const ComponentWithCleanup = () => {

  useEffect(() => {
    console.log('I\'m in useEffect!');
    console.log('This will be called whenever an instance of this component mounts');

    return () => { console.log('useEffect calls the returned function during componentWillUnmount') };
  }, []);

  return (<h1>Demo of Clean-up Function Syntax</h1>);
};

function App() {
  const [showComponent, setShowComponent] = useState(true);

  const childComponent = showComponent ? <ComponentWithCleanup /> : <h1>Removed!</h1>;

  return (
    <div>
      { childComponent }
      <button onClick={() => setShowComponent(! showComponent)}>Update</button>
    </div>
  );
}
```